arch/x86_64/mcount-insn.c

```diff
@@ -193,6 +193,209 @@ static int handle_pic_lea(cs_insn *insn, uint8_t insns[],
 	return -1;
 }
 
+/*
+ *  handle PIC code for MOV instruction.
+ *  this function manipulate the instruction.
+ *
+ *  for examples :
+ *
+ *  case LOAD:
+ *	mov rcx, qword ptr [rip + 0x8f3f85]
+ *  to manipulate like below:
+ *	push rax
+ *	mov rax, PC_BASE + 0x8f3f85
+ *	mov rcx, [rax]
+ *	pop rax
+ *
+ *  case STORE:
+ *  	mov qword ptr [rip+0x8f3f85], rcx
+ *  to manipulate like below:
+ *	push rax
+ *	mov rax, PC_BASE + 0x8f3f85
+ *	mov [rax], rcx
+ *	pop rax
+ *
+ *  exceptional case in LOAD :
+ *  	mov rax, qword ptr [rip+0x8f3f85]
+ *  this will change like below:
+ * 	xchg rax, rcx
+ *  	push rax
+ *  	mov rax, PC_BASE + 0x8f3f85
+ *  	mov rcx, [rax]
+ *  	pop rax
+ *  	xchg rax, rcx
+ *
+ *  exceptional case in STORE :
+ *  	mov qword ptr [rip+0x8f3f85], rax
+ *  to manipulate like below:
+ *  	xchg rax, rcx
+ *	push rax
+ *	mov rax, PC_BASE + 0x8f3f85
+ *	mov [rax], rcx
+ *	pop rax
+ *	xchg rax, rcx
+ *
+ */
+static int handle_pic_mov(cs_insn *insn, uint8_t insns[],
+		      struct mcount_disasm_info *info)
+{
+	struct MODRM modrm;
+	int modrm_rows, modrm_cols;
+	effect_t effect_opnd;
+	uint64_t PC_base;
+
+	cs_x86 *x86 = &insn->detail->x86;
+	cs_x86_op *opnd1 = &x86->operands[0];
+	cs_x86_op *opnd2 = &x86->operands[1];
+
+	static const int NEED_XCHG = 0;
+	static const int NON_XCHG = 2;
+	static const int LOAD = 0;
+	static const int STORE = 1;
+
+	int adjust = NON_XCHG;
+	int insns_type = LOAD;
+
+	// typically instruction elements.
+	uint8_t mov_insns[20] = {0xCC};
+	int insns_size;
+	enum mov_insn_map {
+		XCHG_1 = 0,
+		PUSH = 2,
+		REX_P,
+		OP,
+		IMM,
+		IMM_END = 12,
+		REX_P2 = 13,
+		PREFIX_P2 = 13,
+		OP2,
+		MODRM,
+		POP = 16,
+		XCHG_2,
+	};
+
+	/* support MOV instruction only */
+	if (strcmp(insn->mnemonic, "mov") != 0)
+		goto out;
+
+	/* check PC-relative addressing mode */
+	if (opnd1->type == X86_OP_MEM && opnd1->mem.base == X86_REG_RIP) {
+		insns_type = STORE;
+	}
+	else if (opnd2->type == X86_OP_MEM && opnd2->mem.base == X86_REG_RIP) {
+		insns_type = LOAD;
+	}
+	else
+		goto out;
+
+	/* the SIB addressing is not supported yet */
+	if (opnd1->mem.scale > 1 || opnd2->mem.scale > 1)
+		goto out;
+
+	/* for handle exceptional case, */
+	if (insns_type == LOAD) {
+		/* for now, support only 64bit registere */
```

Comment:
typo,  anyway as far as I know RIP-relative addressing is only available in 64-bit mode.  Do you really need this?  Also it might not be a contiguous range from RAX to R15.

Reply:
@namhyung I'm pretty sure about this. because operand is not taken 64 bit register only, it can be "EAX or AX".